src/StreamJsonRpc.Tests/JsonRpcWithFatalExceptionsTests.cs

```diff
@@ -0,0 +1,321 @@
+    {
+        var exceptionMessage = "Exception from CloseStreamOnAsyncMethodException";
+        await Assert.ThrowsAsync<TaskCanceledException>(() => this.clientRpc.InvokeAsync(nameof(Server.MethodThatThrowsAsync), exceptionMessage));
+        Assert.Equal(FaultException.Message, exceptionMessage);
+
+        // Assert that the JsonRpc, MessageHandler, and Stream objects are disposed after exception
+        await Assert.ThrowsAsync<ObjectDisposedException>(() => this.clientRpc.InvokeAsync(nameof(Server.MethodThatThrowsAsync), exceptionMessage));
+        Assert.True(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
+        Assert.True(this.serverStream.IsDisposed);
+        Assert.True(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    public async Task CloseStreamOnAsyncTMethodException()
+    {
+        var exceptionMessage = "Exception from CloseStreamOnAsyncTMethodException";
+        await Assert.ThrowsAsync<TaskCanceledException>(() => this.clientRpc.InvokeAsync<string>(nameof(Server.AsyncMethodThatReturnsStringAndThrows), exceptionMessage));
+        Assert.Equal(FaultException.Message, exceptionMessage);
+
+        // Assert that the JsonRpc, MessageHandler, and Stream objects are disposed after exception
+        await Assert.ThrowsAsync<ObjectDisposedException>(() => this.clientRpc.InvokeAsync(nameof(Server.AsyncMethodThatReturnsStringAndThrows), exceptionMessage));
+        Assert.True(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
+        Assert.True(this.serverStream.IsDisposed);
+        Assert.True(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    public async Task StreamsStayOpenForNonServerException()
+    {
+        await Assert.ThrowsAsync(typeof(RemoteMethodNotFoundException), () => this.clientRpc.InvokeAsync("missingMethod", 50));
+
+        // Assert MessageHandler and Stream objects are not disposed
+        Assert.False(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
+        Assert.False(this.serverStream.IsDisposed);
+        Assert.False(this.clientStream.IsDisposed);
+    }
+
+    [Fact]
+    public async Task StreamsStayOpenOnOperationCanceled()
+    {
+        using (var cts = new CancellationTokenSource())
+        {
+            var invokeTask = this.clientRpc.InvokeWithCancellationAsync<string>(nameof(Server.AsyncMethodWithCancellation), new[] { "a" }, cts.Token);
+            await this.server.ServerMethodReached.WaitAsync(this.TimeoutToken);
+            cts.Cancel();
+
+            // Ultimately, the server throws because it was canceled.
+            var ex = await Assert.ThrowsAnyAsync<OperationCanceledException>(() => invokeTask.WithTimeout(UnexpectedTimeout));
+#if !NET452
+            Assert.Equal(cts.Token, ex.CancellationToken);
+#endif
+        }
+
+        // Assert MessageHandler and Stream objects are not disposed
+        Assert.False(((DisposingMessageHander)this.clientRpc.MessageHandler).IsDisposed);
```

Comment:
Why shouldn't they be disposed on cancellation? Your `IsFatalException` override returns true unconditionally. So I read that to mean that `TaskCanceledException` is fatal.

Reply:
Good point. I was thinking we only needed to worry about exceptions thrown explicitly from the server method (but this would miss the whole gambit of `OperationCanceledExceptions`. Fixed in my next iteration

Reply:
>  I was thinking we only needed to worry about exceptions thrown explicitly from the server method

I would agree -- but when a server method is canceled, it will in fact throw this exception -- unless perhaps the request to invoke is canceled before we even invoke the method to begin with, in which case the exception might never be thrown, theoretically. And in that case, I guess I'd be ok with us not calling IsFatalException at all in that case since indeed the server method didn't throw it.

Reply:
Regardless, I was missing the case where the server explicitly throws an `OperationCanceledException` in an async method, so it was a great catch!